components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt

```diff
@@ -0,0 +1,231 @@
+                put(MediaStore.Downloads.SIZE, download.contentLength)
+                put(MediaStore.Downloads.IS_PENDING, 1)
+            }
+
+            val resolver = applicationContext.contentResolver
+            val collection = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
+            val item = resolver.insert(collection, values)
+
+            val pfd = resolver.openFileDescriptor(item!!, "w")
+            ParcelFileDescriptor.AutoCloseOutputStream(pfd).use(block)
+
+            values.clear()
+            values.put(MediaStore.Downloads.IS_PENDING, 0)
+            resolver.update(item, values, null, null)
+        } else {*/
+        val dir = Environment.getExternalStoragePublicDirectory(download.destinationDirectory)
+        val file = File(dir, filename)
+        FileOutputStream(file).use(block)
+
+        val contentType = response.headers[CONTENT_TYPE]
+        val contentLength = response.headers[CONTENT_LENGTH]?.toLongOrNull()
+
+        addCompletedDownload(
+            title = filename,
+            description = filename,
+            isMediaScannerScannable = true,
+            mimeType = contentType ?: download.contentType ?: "*/*",
+            path = file.absolutePath,
+            length = contentLength ?: download.contentLength ?: file.length(),
+            showNotification = true, // TODO show when our notifications are blocked
+            uri = download.url.toUri(),
+            referer = download.referrerUrl?.toUri()
+        )
+        // }
+    }
+
+    @TargetApi(Build.VERSION_CODES.P)
+    @Suppress("Deprecation", "LongParameterList", "LongMethod")
+    private fun addCompletedDownload(
+        title: String,
+        description: String,
+        isMediaScannerScannable: Boolean,
+        mimeType: String,
+        path: String,
+        length: Long,
+        showNotification: Boolean,
+        uri: Uri,
+        referer: Uri?
+    ) = getSystemService<SystemDownloadManager>()!!.run {
+        if (SDK_INT >= Build.VERSION_CODES.N) {
+            addCompletedDownload(
+                title,
+                description,
+                isMediaScannerScannable,
+                mimeType,
+                path,
+                length,
+                showNotification,
+                uri,
+                referer
+            )
+        } else {
+            addCompletedDownload(
+                title,
+                description,
+                isMediaScannerScannable,
+                mimeType,
+                path,
+                length,
+                showNotification
+            )
+        }
+    }
+
+    private fun buildNotification(): Notification {
+        val channelId = DownloadNotificationChannel.ensureChannelExists(this)
+
+        return NotificationCompat.Builder(this, channelId)
+            .setSmallIcon(android.R.drawable.stat_sys_download)
+            .setContentTitle("Downloading...")
```

Comment:
nit: Needs to move to strings.xml (and check whether that file is listed in l10n.toml.